File: android/src/main/kotlin/com/karnival/barcode_scanner_pro/CameraManager.kt
package com.karnival.barcode_scanner_pro

import android.content.Context
import android.graphics.Bitmap
import android.util.Size
import androidx.camera.core.Camera
import androidx.camera.core.CameraSelector
import androidx.camera.core.FocusMeteringAction
import androidx.camera.core.ImageAnalysis
import androidx.camera.core.Preview
import androidx.camera.core.resolutionselector.ResolutionSelector
import androidx.camera.core.resolutionselector.ResolutionStrategy
import androidx.camera.lifecycle.ProcessCameraProvider
import androidx.camera.view.PreviewView
import androidx.core.content.ContextCompat
import androidx.lifecycle.LifecycleOwner
import java.io.ByteArrayOutputStream
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit

/**
 * Owns the CameraX pipeline for one scanner view:
 *
 *   ProcessCameraProvider -> Preview -> ImageAnalysis -> background Executor.
 *
 * Camera control (torch/zoom/exposure/focus) is delegated to [Camera.getCameraControl].
 * Binding happens on the main thread; analysis runs on a dedicated single-thread
 * executor so the UI thread is never blocked. `KEEP_ONLY_LATEST` backpressure
 * guarantees we always analyze the most recent frame and recycle the rest.
 */
internal class CameraManager(
    private val context: Context,
    private val previewView: PreviewView,
    private val lifecycleOwner: LifecycleOwner,
    private val analyzer: FrameAnalyzer,
    private val events: EventDispatcher,
) {
    private val analysisExecutor: ExecutorService = Executors.newSingleThreadExecutor()
    private var cameraProvider: ProcessCameraProvider? = null
    private var camera: Camera? = null
    private var imageAnalysis: ImageAnalysis? = null
    private var preview: Preview? = null

    private var lensFacing = CameraSelector.LENS_FACING_BACK

    /** Starts the camera with [config]. [onReady]/[onError] are invoked on main. */
    fun start(config: ScannerConfig, onReady: () -> Unit, onError: (String, String) -> Unit) {
        lensFacing = config.lensFacing
        val future = ProcessCameraProvider.getInstance(context)
        future.addListener({
            try {
                cameraProvider = future.get()
                bindUseCases(config)
                onReady()
            } catch (e: Exception) {
                onError(ErrorCode.INITIALIZATION_FAILED, e.localizedMessage ?: "Camera init failed")
            }
        }, ContextCompat.getMainExecutor(context))
    }

    private fun bindUseCases(config: ScannerConfig) {
        val provider = cameraProvider ?: return
        val res = config.targetResolution
        val resolutionSelector = ResolutionSelector.Builder()
            .setResolutionStrategy(
                ResolutionStrategy(Size(res.width, res.height),
                    ResolutionStrategy.FALLBACK_RULE_CLOSEST_HIGHER_THEN_LOWER),
            )
            .build()

        preview = Preview.Builder().build().also {
            it.setSurfaceProvider(previewView.surfaceProvider)
        }

        imageAnalysis = ImageAnalysis.Builder()
            .setResolutionSelector(resolutionSelector)
            .setBackpressureStrategy(ImageAnalysis.STRATEGY_KEEP_ONLY_LATEST)
            .setOutputImageFormat(ImageAnalysis.OUTPUT_IMAGE_FORMAT_YUV_420_888)
            .build()
            .also { it.setAnalyzer(analysisExecutor, analyzer) }

        val selector = CameraSelector.Builder().requireLensFacing(lensFacing).build()
        provider.unbindAll()
        camera = provider.bindToLifecycle(lifecycleOwner, selector, preview, imageAnalysis)
    }

    fun switchCamera(config: ScannerConfig) {
        lensFacing = if (lensFacing == CameraSelector.LENS_FACING_BACK) {
            CameraSelector.LENS_FACING_FRONT
        } else {
            CameraSelector.LENS_FACING_BACK
        }
        bindUseCases(config)
        events.send(
            mapOf(
                "type" to EventType.CAMERA_CHANGED,
                "camera" to if (lensFacing == CameraSelector.LENS_FACING_FRONT) 1 else 0,
            ),
        )
    }

    // --- Controls --------------------------------------------------------------

    fun setTorch(enabled: Boolean) {
        camera?.cameraControl?.enableTorch(enabled)
        events.send(mapOf("type" to EventType.FLASH_CHANGED, "enabled" to enabled))
    }

    fun toggleTorch(): Boolean {
        val on = (camera?.cameraInfo?.torchState?.value ?: 0) == 1
        setTorch(!on)
        return !on
    }

    /** [value] normalized in `[0,1]`. */
    fun setZoom(value: Float) {
        camera?.cameraControl?.setLinearZoom(value.coerceIn(0f, 1f))
        events.send(mapOf("type" to EventType.ZOOM_CHANGED, "zoom" to value.toDouble()))
    }

    /** [value] in `[-1,1]` mapped onto the device EV range. */
    fun setExposure(value: Float) {
        val info = camera?.cameraInfo ?: return
        val range = info.exposureState.exposureCompensationRange
        val target = if (value >= 0) (value * range.upper) else (-value * range.lower)
        camera?.cameraControl?.setExposureCompensationIndex(target.toInt())
    }

    /** [nx]/[ny] normalized `[0,1]` in preview space. */
    fun focusAt(nx: Float, ny: Float) {
        val factory = previewView.meteringPointFactory
        val point = factory.createPoint(nx * previewView.width, ny * previewView.height)
        val action = FocusMeteringAction.Builder(point)
            .setAutoCancelDuration(3, TimeUnit.SECONDS)
            .build()
        camera?.cameraControl?.startFocusAndMetering(action)
    }

    /** Captures the current preview as JPEG bytes (best-effort). */
    fun captureFrame(): ByteArray? {
        val bitmap: Bitmap = previewView.bitmap ?: return null
        return ByteArrayOutputStream().use { out ->
            bitmap.compress(Bitmap.CompressFormat.JPEG, 85, out)
            out.toByteArray()
        }
    }

    fun pauseAnalysis() {
        analyzer.active = false
    }

    fun resumeAnalysis() {
        analyzer.resetDuplicates()
        analyzer.active = true
    }

    fun stop() {
        analyzer.active = false
        cameraProvider?.unbindAll()
    }

    fun dispose() {
        stop()
        analysisExecutor.shutdown()
        camera = null
        cameraProvider = null
    }
}

File: ios/Classes/CameraManager.swift
import AVFoundation
import CoreImage
import ImageIO
import UIKit

/// Owns the AVFoundation capture pipeline for one scanner view:
///
///   AVCaptureSession -> AVCaptureVideoDataOutput -> CMSampleBuffer -> Vision.
///
/// Session configuration and control run on a dedicated serial queue so the
/// main thread is never blocked. The most recent pixel buffer is retained for
/// `captureFrame`. Camera controls are delegated to the active `AVCaptureDevice`.
final class CameraManager: NSObject, AVCaptureVideoDataOutputSampleBufferDelegate {

  private let session = AVCaptureSession()
  private let sessionQueue = DispatchQueue(label: "com.umda.barcode_scanner_pro.session")
  private let videoOutput = AVCaptureVideoDataOutput()
  private let analyzer: FrameAnalyzer
  private let events: EventDispatcher
  private let ciContext = CIContext()

  private var device: AVCaptureDevice?
  private var position: AVCaptureDevice.Position = .back
  private var latestPixelBuffer: CVPixelBuffer?

  let previewLayer: AVCaptureVideoPreviewLayer

  init(analyzer: FrameAnalyzer, events: EventDispatcher) {
    self.analyzer = analyzer
    self.events = events
    self.previewLayer = AVCaptureVideoPreviewLayer(session: session)
    super.init()
    self.previewLayer.videoGravity = .resizeAspectFill
    NotificationCenter.default.addObserver(
      self, selector: #selector(orientationChanged),
      name: UIDevice.orientationDidChangeNotification, object: nil)
    updateOrientation()
  }

  // MARK: - Lifecycle

  func start(
    config: ScannerConfig, onReady: @escaping () -> Void,
    onError: @escaping (String, String) -> Void
  ) {
    position = config.devicePosition
    sessionQueue.async { [weak self] in
      guard let self = self else { return }
      do {
        try self.configureSession(config)
        self.session.startRunning()
        DispatchQueue.main.async(execute: onReady)
      } catch let e as ConfigError {
        DispatchQueue.main.async { onError(e.code, e.message) }
      } catch {
        DispatchQueue.main.async {
          onError(ErrorCode.initializationFailed, error.localizedDescription)
        }
      }
    }
  }

  private struct ConfigError: Error { let code: String; let message: String }

  private func configureSession(_ config: ScannerConfig) throws {
    session.beginConfiguration()
    defer { session.commitConfiguration() }

    if session.canSetSessionPreset(config.sessionPreset) {
      session.sessionPreset = config.sessionPreset
    }

    // Remove any previous I/O (e.g. on camera switch).
    session.inputs.forEach { session.removeInput($0) }
    session.outputs.forEach { session.removeOutput($0) }

    guard let camera = Self.device(for: position) else {
      throw ConfigError(code: ErrorCode.cameraUnavailable, message: "No camera for position")
    }
    device = camera
    let input = try AVCaptureDeviceInput(device: camera)
    guard session.canAddInput(input) else {
      throw ConfigError(code: ErrorCode.initializationFailed, message: "Cannot add camera input")
    }
    session.addInput(input)

    videoOutput.alwaysDiscardsLateVideoFrames = true
    videoOutput.setSampleBufferDelegate(self, queue: sessionQueue)
    guard session.canAddOutput(videoOutput) else {
      throw ConfigError(code: ErrorCode.initializationFailed, message: "Cannot add video output")
    }
    session.addOutput(videoOutput)

    if config.enableAutoFocus, camera.isFocusModeSupported(.continuousAutoFocus) {
      try? camera.lockForConfiguration()
      camera.focusMode = .continuousAutoFocus
      camera.unlockForConfiguration()
    }
  }

  private static func device(for position: AVCaptureDevice.Position) -> AVCaptureDevice? {
    AVCaptureDevice.DiscoverySession(
      deviceTypes: [.builtInWideAngleCamera],
      mediaType: .video, position: position
    ).devices.first
  }

  func switchCamera(config: ScannerConfig) {
    position = (position == .back) ? .front : .back
    sessionQueue.async { [weak self] in
      guard let self = self else { return }
      try? self.configureSession(config)
      self.events.send([
        "type": EventType.cameraChanged,
        "camera": self.position == .front ? 1 : 0,
      ])
    }
  }

  func pauseAnalysis() { analyzer.active = false }

  func resumeAnalysis() {
    analyzer.resetDuplicates()
    analyzer.active = true
  }

  func stop() {
    analyzer.active = false
    sessionQueue.async { [weak self] in self?.session.stopRunning() }
  }

  func dispose() {
    NotificationCenter.default.removeObserver(self)
    stop()
    sessionQueue.async { [weak self] in
      self?.videoOutput.setSampleBufferDelegate(nil, queue: nil)
      self?.latestPixelBuffer = nil
    }
  }

  // MARK: - Controls

  func setTorch(_ on: Bool) {
    sessionQueue.async { [weak self] in
      guard let d = self?.device, d.hasTorch else { return }
      try? d.lockForConfiguration()
      d.torchMode = on ? .on : .off
      d.unlockForConfiguration()
      self?.events.send(["type": EventType.flashChanged, "enabled": on])
    }
  }

  func toggleTorch() -> Bool {
    let on = device?.torchMode == .on
    setTorch(!on)
    return !on
  }

  /// [value] normalized in `[0,1]`.
  func setZoom(_ value: Double) {
    sessionQueue.async { [weak self] in
      guard let d = self?.device else { return }
      let maxFactor = min(d.activeFormat.videoMaxZoomFactor, 8.0)
      let factor = 1.0 + CGFloat(max(0, min(1, value))) * (maxFactor - 1.0)
      try? d.lockForConfiguration()
      d.videoZoomFactor = factor
      d.unlockForConfiguration()
      self?.events.send(["type": EventType.zoomChanged, "zoom": value])
    }
  }

  /// [value] in `[-1,1]` mapped onto the device exposure-bias range.
  func setExposure(_ value: Double) {
    sessionQueue.async { [weak self] in
      guard let d = self?.device else { return }
      let bias: Float =
        value >= 0
        ? Float(value) * d.maxExposureTargetBias
        : Float(-value) * d.minExposureTargetBias
      try? d.lockForConfiguration()
      d.setExposureTargetBias(bias, completionHandler: nil)
      d.unlockForConfiguration()
    }
  }

  /// [nx]/[ny] normalized `[0,1]` in preview space. Conversion to device space
  /// uses the preview layer and therefore runs on the main thread first.
  func focusAt(nx: Double, ny: Double) {
    let layerPoint = CGPoint(
      x: nx * previewLayer.bounds.width, y: ny * previewLayer.bounds.height)
    let devicePoint = previewLayer.captureDevicePointConverted(fromLayerPoint: layerPoint)
    sessionQueue.async { [weak self] in
      guard let d = self?.device else { return }
      try? d.lockForConfiguration()
      if d.isFocusPointOfInterestSupported {
        d.focusPointOfInterest = devicePoint
        d.focusMode = .autoFocus
      }
      if d.isExposurePointOfInterestSupported {
        d.exposurePointOfInterest = devicePoint
        d.exposureMode = .autoExpose
      }
      d.unlockForConfiguration()
    }
  }

  /// Captures the latest frame as JPEG bytes (best-effort).
  func captureFrame() -> Data? {
    guard let pb = latestPixelBuffer else { return nil }
    let ci = CIImage(cvPixelBuffer: pb)
    guard let cg = ciContext.createCGImage(ci, from: ci.extent) else { return nil }
    return UIImage(cgImage: cg).jpegData(compressionQuality: 0.85)
  }

  // MARK: - Orientation

  @objc private func orientationChanged() { updateOrientation() }

  private func updateOrientation() {
    let orientation = UIDevice.current.orientation
    let cg: CGImagePropertyOrientation
    let video: AVCaptureVideoOrientation
    switch orientation {
    case .landscapeLeft: cg = .up; video = .landscapeRight
    case .landscapeRight: cg = .down; video = .landscapeLeft
    case .portraitUpsideDown: cg = .left; video = .portraitUpsideDown
    default: cg = .right; video = .portrait
    }
    analyzer.orientation = cg
    if let conn = previewLayer.connection, conn.isVideoOrientationSupported {
      conn.videoOrientation = video
    }
  }

  // MARK: - Sample buffer delegate

  func captureOutput(
    _ output: AVCaptureOutput, didOutput sampleBuffer: CMSampleBuffer,
    from connection: AVCaptureConnection
  ) {
    latestPixelBuffer = CMSampleBufferGetImageBuffer(sampleBuffer)
    analyzer.analyze(sampleBuffer)
  }
}
